Initial support for datetimeDiff

dpsutton opened the pull request:
[Epic](https://github.com/metabase/metabase/issues/19274)

Introduces a notion of `datetimediff` to mbql. `datetimediff` takes two temporal expressions and a unit and returns the number of units between the two expressions.

```
datetimediff( temporal-expression-1, temporal-expression-2, unit )

unit : year |  month | week | day | hour | minute | second

The result is positive if temporal-expression-1 <= temporal-expression-2, and negative otherwise.

Days, weeks, months, and years are only counted if they are whole to the "day". For example, 
`datetimediff("2022-01-30", "2022-02-28", "month")` returns 0 months. 
If the values are timestamps, the time of the day doesn't matter to the calculation. For example, 
`datetimediff("2022-01-01T09:00:00", "2022-01-02T08:00:00", "day")` returns 1 day even though it is less than 24 hours.

Similarly, hours, minutes, and seconds are only counted if they are whole. 
For example, datetimediff("2022-01-00T01:00:30", "2022-01-01T02:00:29", "hour") returns 0 hours.
```

Docs in the GUI editor:
<img width="410" alt="image" src="https://user-images.githubusercontent.com/39073188/198113473-b39e3858-f04a-464a-8e0b-bc7c8dc2b720.png">

Initially this is supported for the following drivers:
- mysql
- postgres
- bigquery-cloud-sdk

Using the sample dataset on _POSTGRES_ I'm joining orders to reviews by the product id.
> :warning: **NOTE:** screenshots from when the function was called `datetime` but is now renamed to `datetimediff`

A custom column for the difference in days between the review and order creation
![image](https://user-images.githubusercontent.com/39073188/198226948-656a2a2b-f9dd-4507-8f95-fe64b79308b6.png)

and filtering where the review was created within 2 months of the order
<img width="875" alt="image" src="https://user-images.githubusercontent.com/6377293/194935628-55390ec3-41f5-47ae-9afe-2b64171fe34d.png">


<img width="1282" alt="image" src="https://user-images.githubusercontent.com/6377293/194935652-d80cab3a-060d-4572-9210-3fcbdbdc1cdc.png">


<img width="1282" alt="image" src="https://user-images.githubusercontent.com/6377293/194935673-2162eb83-d010-4306-9d0a-954864c11748.png">

dpsutton committed:
Initial support for pg for `date-diff`

```clojure
dates=> (metabase.test/rows
          (qp/process-query
            {:type :query,
             :query {:source-table 84,
                     :joins [{:fields [],
                              :source-table 85,
                              :condition [:=
                                          [:field 650 nil]
                                          [:field 660 {:join-alias "Reviews"}]],
                              :alias "Reviews"}],
                     :fields [[:field 648 nil]
                              [:field 657 {:join-alias "Reviews"}]
                              [:expression "diff-day"]
                              [:expression "diff-month"]],
                     :expressions {"diff-day" [:datediff
                                               ["field" 657 {"join-alias" "Reviews"}]
                                               ["field" 648 nil]
                                               :day]
                                   "diff-month" [:datediff
                                                 ["field" 657 {"join-alias" "Reviews"}]
                                                 ["field" 648 nil]
                                                 :month]}
                     :limit 10}
             :database 3}))
[["2017-07-19T19:44:56.582Z" "2018-05-15T20:25:48.517Z" 300.0 9.0]
 ["2017-07-19T19:44:56.582Z" "2019-08-07T13:50:33.401Z" 748.0 24.0]
 ["2017-07-19T19:44:56.582Z" "2018-03-30T00:28:45.192Z" 253.0 8.0]
 ["2017-07-19T19:44:56.582Z" "2017-11-13T22:29:12.121Z" 117.0 3.0]
 ["2017-07-19T19:44:56.582Z" "2017-11-19T07:08:54.771Z" 122.0 3.0]
 ["2017-07-19T19:44:56.582Z" "2018-02-11T03:05:17.346Z" 206.0 6.0]
 ["2017-07-19T19:44:56.582Z" "2020-01-31T12:35:59.147Z" 925.0 30.0]
 ["2017-07-19T19:44:56.582Z" "2019-12-10T16:16:36.999Z" 873.0 28.0]
 ["2018-09-08T22:03:20.239Z" "2019-11-06T10:43:14.868Z" 423.0 13.0]
 ["2018-09-08T22:03:20.239Z" "2019-02-13T17:37:35.244Z" 157.0 5.0]]

dates=> (metabase.test/rows
          (qp/process-query
            {:type :query,
             :query {:source-table 84,
                     :joins [{:fields [[:field 657 {:join-alias "Reviews"}]],
                              :source-table 85,
                              :condition [:=
                                          [:field 650 nil]
                                          [:field 660 {:join-alias "Reviews"}]],
                              :alias "Reviews"}],
                     :fields [[:field 648 nil]
                              [:expression "diff-year"]
                              [:expression "diff-month"]
                              [:expression "diff-day"]
                              [:expression "diff-hour"]
                              [:expression "diff-minute"]
                              [:expression "diff-second"]],
                     :expressions {"diff-year" [:datediff
                                                [:field 657 {:join-alias "Reviews"}]
                                                [:field 648 nil]
                                                :year]
                                   "diff-month" [:datediff
                                                 [:field 657 {:join-alias "Reviews"}]
                                                 [:field 648 nil]
                                                 :month]
                                   "diff-day" [:datediff
                                               [:field 657 {:join-alias "Reviews"}]
                                               [:field 648 nil]
                                               :day]
                                   "diff-hour" [:datediff
                                                [:field 657 {:join-alias "Reviews"}]
                                                [:field 648 nil]
                                                :hour]
                                   "diff-minute" [:datediff
                                                  [:field 657 {:join-alias "Reviews"}]
                                                  [:field 648 nil]
                                                  :minute]
                                   "diff-second" [:datediff
                                                  [:field 657 {:join-alias "Reviews"}]
                                                  [:field 648 nil]
                                                  :second]}
                     :limit 10},
             :database 3}))
[["2017-07-19T19:44:56.582Z"
  0.0
  9.0
  300.0
  7200.0
  432040.0
  2.5922451935E7
  "2018-05-15T20:25:48.517Z"]
 ["2017-07-19T19:44:56.582Z"
  2.0
  24.0
  748.0
  17970.0
  1078205.0
  6.4692336819E7
  "2019-08-07T13:50:33.401Z"]
 ["2017-07-19T19:44:56.582Z"
  0.0
  8.0
  253.0
  6076.0
  364603.0
  2.187622861E7
  "2018-03-30T00:28:45.192Z"]
 ["2017-07-19T19:44:56.582Z"
  0.0
  3.0
  117.0
  2810.0
  168644.0
  1.0118655539E7
  "2017-11-13T22:29:12.121Z"]
 ["2017-07-19T19:44:56.582Z"
  0.0
  3.0
  122.0
  2939.0
  176363.0
  1.0581838189E7
  "2017-11-19T07:08:54.771Z"]
 ["2017-07-19T19:44:56.582Z"
  0.0
  6.0
  206.0
  4951.0
  297080.0
  1.7824820764E7
  "2018-02-11T03:05:17.346Z"]
 ["2017-07-19T19:44:56.582Z"
  2.0
  30.0
  925.0
  22216.0
  1333011.0
  7.9980662565E7
  "2020-01-31T12:35:59.147Z"]
 ["2017-07-19T19:44:56.582Z"
  2.0
  28.0
  873.0
  20972.0
  1258351.0
  7.5501100417E7
  "2019-12-10T16:16:36.999Z"]
 ["2018-09-08T22:03:20.239Z"
  1.0
  13.0
  423.0
  10164.0
  609879.0
  3.6592794629E7
  "2019-11-06T10:43:14.868Z"]
 ["2018-09-08T22:03:20.239Z"
  0.0
  5.0
  157.0
  3787.0
  227254.0
  1.3635255005E7
  "2019-02-13T17:37:35.244Z"]]
```

dpsutton committed:
bigquery datediff

dpsutton committed:
Ensure pg datediff results are integral

dpsutton committed:
Typo in bigquery driver name

dpsutton committed:
Tests for datediff

dpsutton committed:
mysql datediff

args for the datediff are intentionally swapped as it does a subtraction
and we want to align it with the others.

ignoring the dates returned since mysql seems to have an issue with
timezones with the sample dataset

dpsutton committed:
Tests for edge cases

made a new .edn dataset definition. Expectation is that we continue to
add more stuff into there rather than make one-off `mt/dataset`
defintions in random test files.

Why? Because bigquery-cloud-sdk differs slightly from mysql/pg in the
number of days and years on some edge cases. And rather than putting
edge cases inside of a single test that runs all of the queries and
cases on the `driver/*driver*`, we can just let each one refer to the
test dataset and assert what it wants.

Before:

```clojure
(is (= (case driver/*driver :bigquery-cloud-sdk [1 365] [0 364])
       (mt/rows (mt/run-mbql-query ...))))
```

After, each driver namespace can test against those values as it
likes. This lets third party drivers also test against this. they would
have to add their own overrides to the main repo otherwise.

dpsutton committed:
Test that datediff can compare across all temporal types

dpsutton committed:
Verify types in bigquery

dpsutton committed:
Redshift does not have a time column so remove the time table

dpsutton committed:
bigquery tests against time column dataset

dpsutton committed:
Disable datediff from redshift for the moment (for ever?)

redshift doesn't have `date_part(unit, interval)` that we use in
postgres. It does have a `datediff` function but that tracks the
crossing of boundaries. From its documentation:

> Specifically, DATEDIFF determines the number of date part boundaries
that are crossed between two expressions. For example, suppose that
you're calculating the difference in years between two dates, 12-31-2008
and 01-01-2009. In this case, the function returns 1 year despite the
fact that these dates are only one day apart. If you are finding the
difference in hours between two timestamps, 01-01-2009 8:30:00 and
01-01-2009 10:00:00, the result is 2 hours. If you are finding the
difference in hours between two timestamps, 8:30:00 and 10:00:00, the
result is 2 hours.

8:30 to 10:00 is 2 hours, 12-31 and 1-1 is one year despite being one
day apart.

dpsutton committed:
Disable datediff from redshift for the moment (for ever?)

redshift doesn't have `date_part(unit, interval)` that we use in
postgres. It does have a `datediff` function but that tracks the
crossing of boundaries. From its documentation:

> Specifically, DATEDIFF determines the number of date part boundaries
that are crossed between two expressions. For example, suppose that
you're calculating the difference in years between two dates, 12-31-2008
and 01-01-2009. In this case, the function returns 1 year despite the
fact that these dates are only one day apart. If you are finding the
difference in hours between two timestamps, 01-01-2009 8:30:00 and
01-01-2009 10:00:00, the result is 2 hours. If you are finding the
difference in hours between two timestamps, 8:30:00 and 10:00:00, the
result is 2 hours.

8:30 to 10:00 is 2 hours, 12-31 and 1-1 is one year despite being one
day apart.

dpsutton committed:
turn on bigquery fks in tests

Running tests against bigquery locally in my running repl passed. But
they failed in CI saying that left joins are not supported.

```
check_features.clj:14
:bigquery-cloud-sdk
using sample-dataset dataset
Computes without errors
expected: (= [[0 0 1 39 2384 143082] [0 0 1 30 1856 111388]] (mt/rows (qp/process-query query)))
  actual: #error {
 :cause "left-join is not supported by this driver."
 :data {:type :unsupported-feature, :feature :left-join}
 :via
 [{:type clojure.lang.ExceptionInfo
   :message "left-join is not supported by this driver."
   :data {:type :unsupported-feature, :feature :left-join}
   :at [metabase.query_processor.middleware.check_features$assert_driver_supports invokeStatic "check_features.clj" 14]}]
```

In mb/driver/sql.clj:

```clojure
(doseq [join-feature [:left-join
                      :right-join
                      :inner-join
                      :full-join]]
  (defmethod driver/supports? [:sql join-feature]
    [driver _]
    (driver/supports? driver :foreign-keys)))
```

And the bigquery driver:

```clojure
(defmethod driver/supports? [:bigquery-cloud-sdk :foreign-keys] [_ _] true)
```

So what gives? In the test data bigquery namespace:

```clojure
;; Don't enable foreign keys when testing because BigQuery *doesn't* have a notion of foreign keys. Joins are still
;; allowed, which puts us in a weird position, however; people can manually specifiy "foreign key" relationships in
;; admin and everything should work correctly. Since we can't infer any "FK" relationships during sync our normal FK
;; tests are not appropriate for BigQuery, so they're disabled for the time being.
;;
;; TODO - either write BigQuery-speciifc tests for FK functionality or add additional code to manually set up these FK
;; relationships for FK tables
(defmethod driver/supports? [:bigquery-cloud-sdk :foreign-keys] [_ _] (not config/is-test?))
```

So we can enable with `mt/with-bigquery-fks :bigquery-cloud-sdk`. Future
work will fix this once and for all (we aren't syncing foreign key
relationships in test and that function installs them for test-data and
sample-dataset.

```clojure
(if (str/includes? (:name (data/db)) "sample")
  {[:orders :product_id]  [:products :id]
   [:orders :user_id]     [:people :id]
   [:reviews :product_id] [:products :id]}
  {[:checkins :user_id]   [:users :id]
   [:checkins :venue_id]  [:venues :id]
   [:venues :category_id] [:categories :id]})
```

dpsutton committed:
Make the useful-dates closer to one day

dpsutton committed:
Add simple FE stuff

what this is lacking:
- type inferencer stuff
- doesn't seem to be aware of which dbs to offer these functions

for instance, it offers datediff on h2 but it is only supported on
bigquery-cloud-sdk, postgres, and mysql.

dpsutton committed:
Add feature requirement to datediff (so h2 doesn't offer it)

only supported for mysql, bigquery-cloud-sdk, postgres at the momement
and this is the mechanism to restrict it to the drivers that support
it.

dpsutton committed:
shorter test bodies

dpsutton committed:
Ensure we can use datediff functions in arithmetic expressions

- when checking types, we need to check after a call to
sql.qp/->honeysql (we don't type the :date-add clause but the resulting
interval record
- bigquery doesn't seem to allow an interval to use an expression for
the interval amount. Not sure if this a limitation of our side or
bigquery itself

dpsutton committed:
Correctly disable datediff for redshift

dpsutton committed:
simplify var names

dpsutton committed:
Support week

dpsutton commented on modules/drivers/bigquery-cloud-sdk/src/metabase/driver/bigquery_cloud_sdk/query_processor.clj:
```diff
@@ -519,6 +519,49 @@
     (cond->> ((get-method sql.qp/->honeysql [:sql :relative-datetime]) driver clause)
       t (->temporal-type t))))
 
+(defmethod sql.qp/->honeysql [:bigquery-cloud-sdk :datediff]
```
Bigquery is special case for `:week` and others fall through to `date_diff`, `datetime_diff` or `timestamp_diff`. One complication is that `timestamp_diff` does not work for year, month, or week so those fall back to datetime_diff.

dpsutton commented on shared/src/metabase/mbql/schema.cljc:
```diff
@@ -460,7 +466,8 @@
   "Set of valid arithmetic expression clause keywords."
-  #{:+ :- :/ :* :coalesce :length :round :ceil :floor :abs :power :sqrt :log :exp :case})
+  #{:+ :- :/ :* :coalesce :length :round :ceil :floor :abs :power :sqrt :log :exp :case
+    :datediff})
```
I'm not 100% sure of this. Is `arithmetic-expressions` things that operate on numbers, or things that _return_ numbers. I'm working on the latter. So `datediff` returns a number and is fine to use with any of the other arithmetic expressions.

calherries replied:
They are things that return numbers, so this is correct. See for example its use in NumericExpressionArg [here](https://github.com/metabase/metabase/blob/9141582508170bdbe99c4aa74bc4444e41475be6/shared/src/metabase/mbql/schema.cljc#L505).

dpsutton commented on shared/src/metabase/mbql/util.cljc:
```diff
@@ -455,7 +455,10 @@
   [clause]
   (mbql.match/match-one clause
-    #{:interval :relative-datetime}
+    #{:datediff :temporal-extract}
```
This function is used to determine what `:+` is supposed to do in the query processor, either addition time to a date or the more traditional addition of numbers. But now we have datetimes that appear in arithmetic expressions `[:+ 3 [:datediff $created_at $shipped_at :hours]]` is a valid addition expression and should not identify the `{:temporal-unit :default}` of the timestamp columns to mean the whole thing is adding dates together.

calherries replied:
It's also used here in infer-expression-type, to identify expressions that return DateTime types. `https://github.com/metabase/metabase/blob/16ed4b0837580e21126006a06ec57d167d46b654/src/metabase/query_processor/middleware/annotate.clj#L162`

dpsutton commented on src/metabase/driver/mysql.clj:
```diff
@@ -343,6 +347,14 @@
                                 2)
                           (hx/literal "-01"))))
 
+(defmethod sql.qp/->honeysql [:mysql :datediff]
```
mysql is far simpler and we reuse existing function but swap the args 
`[:datediff x y unit]` -> `timestampdiff(y, x, unit)` in mysql

dpsutton commented on src/metabase/driver/postgres.clj:
```diff
@@ -296,6 +300,44 @@
   [driver [_ arg]]
   (sql.qp/->honeysql driver [:percentile arg 0.5]))
 
+(defmethod sql.qp/->honeysql [:postgres :datediff]
```
postgres works the following:
- year: `date_part( year, age(x, y))`
- week: `datediff(x, y, day) / 7`
- month: `datediff(x, y, year) * 12 + date_part( month, age(x, y))`
- day: `date_part( day, x - y)`
- hour: `datediff(x, y, day) * 24 + date_part( hour, age(x, y))`
- minute, and second are similar to hour and month where they get the datediff for the coarser resolution, multiply by 60, and then get the date_part from the age.

dpsutton commented on test/metabase/test/data/dataset_definitions/useful-dates.edn:
```diff
@@ -0,0 +1,18 @@
+;; useful dates for driver tests
```
I made the edn file here so we can let this grow and 3rd party drivers can use these dates as well.

dpsutton committed:
cleanup test

qnkhuat commented on frontend/src/metabase/lib/expressions/helper-text-strings.ts:
```diff
@@ -380,6 +380,26 @@ const helperTextStrings: HelpText[] = [
+      t`month` +
+      ")",
+    args: [
+      {
```
I'd ping @ nllho to have a more detailed docs for the expression.

dpsutton committed:
:datediff -> :datetimediff

qnkhuat commented on modules/drivers/redshift/src/metabase/driver/redshift.clj:
```diff
@@ -24,6 +24,14 @@
 
+(defmethod driver/database-supports? [:redshift :datetimediff]
+  [_driver _feat _db]
+  ;; postgres uses `date_part` on an interval or a call to `age` to get datetimediffs. It seems postgres does not have
```
```suggestion
  ;; postgres uses `date_part` on an interval or a call to `age` to get datetimediffs. It seems redshift does not have
```
I think you mean redshift?

qnkhuat commented on src/metabase/driver/postgres.clj:
```diff
@@ -296,6 +300,44 @@
+    (:year :day)
+    (hx/cast
+     :integer
+     (hsql/call :date_part (hsql/raw (format "'%s'" (name unit)))
```
```suggestion
     (hsql/call :date_part (hx/literal unit)
```

this should have the same effect

qnkhuat reviewed (commented):
I believe we're going with the "whole interval" notion of datediff?


I have some comments regarding tests:
1. One test in `datetimediff-test` uses the sample-dataset test, is there any particular reason we need to use the sample-dataset? Can we use the custom-dataset you have instead? One benefit of using the customdataset is we could easily find the value in the `edn` file and do math with them, versus using the sampledatset we have to join then filter it out.
2. Can you share why you have the edge cases for each driver in different files? I think since we want the result between drivers to behave the same for all drivers, having that test in one file and using `mt/normal-drivers-with-feature` would keep the behavior in-sync and also avoid duplicated tests.
3. For the edge-cases tests, should we test for other units too? (currently we have year and day)

dpsutton committed:
ngoc's suggestions

dpsutton committed:
Better acceptance test for datetimediff

dpsutton committed:
sort ns

dpsutton committed:
embrace the different cases for results

bigquery's documentation says that it returns the number of intervals
but in practice it just returns the number of boundaries

> Returns the whole number of specified date_part intervals between two
DATE objects (date_expression_a - date_expression_b). If the first DATE
is earlier than the second one, the output is negative.

`select date_diff('2022-01-01', '2021-12-31', year)` -> 1

metamben commented on test/metabase/query_processor_test/date_time_zone_functions_test.clj:
```diff
@@ -256,3 +257,46 @@
+                                  :fields [[:expression ~(str "diff-" (name unit))]]
+                                  :expressions
+                                  {~(str "diff-" (name unit))
+                                   [:datetimediff ~'$end ~'$start ~unit]}})))]
```
The macro seems to evaluate `unit` multiple time which could lead to surprises.

Commit:
bigquery day month year

Commit:
Reverse args

Commit:
Update test

Commit:
Centralize tests

Commit:
Change postgres day, month, year behaviour

Commit:
Refactor keep identity

Commit:
Tidy tests

Commit:
Tidy

Commit:
Fix bigquery week

Commit:
Add week tests

Commit:
Fix bigquery week

Commit:
Change mysql day, month, year behaviour

Commit:
Add test for hour, minute, second

dpsutton commented on src/metabase/driver/postgres.clj:
```diff
@@ -296,6 +300,44 @@
+                   (hsql/call :date_part (hsql/raw (format "'%s'" (name unit)))
+                              (hsql/call :age
+                                         (sql.qp/->honeysql driver x)
+                                         (sql.qp/->honeysql driver y)))))))
```
That's a great point. I think we're pivoting to arithmetic based expressions like that for the sub-day intervals which have constant length. (year, month, and day can all be variable)

Commit:
Fix postgres hour minute second

Commit:
Formatting

Commit:
Fix bigquery hour, minute, second

Commit:
Formatting

Commit:
Fix postgres timestamptz

Commit:
WIP

dpsutton committed:
Allow literals in datediff clauses

NOTE: doesn't necessarily allow string literals yet. But makes testing
far easier:

```clojure
date-time-zone-functions-test=>
(mt/rows
 (mt/test-driver :bigquery-cloud-sdk
   (mt/dataset sample-dataset
     (mt/run-mbql-query orders
                        {:limit 1
                         :expressions {"diff"
                                       [:datetimediff #t "2021-10-03" "2022-10-02" :year]}
                         :fields [[:expression "diff"]]}))))
[[0]]
```

calherries committed:
Merge branch 'master' into datediff-functions

Commit:
Uncomment tests

Commit:
Fix bigquery when reporting timezone is not UTC

Commit:
Linting

dpsutton committed:
Moving away from dataset based tests

Commit:
Add timezone tests for week and tidy

Commit:
Remove unused import

Commit:
Consolidate tests

Commit:
Remove with-time-column dataset

Commit:
Remove more-useful-dates defdataset

Commit:
Remove redshift driver WIP

Commit:
Typo

dpsutton committed:
Move DatetimeLiteral clause into DateTimeExpressionArg

The two were separate and not connected.

```clojure
(def ^:private DatetimeLiteral
  "Schema for valid absolute datetime literals."
  (s/conditional
   (partial is-clause? :absolute-datetime)
   absolute-datetime

   (partial is-clause? :time)
   time

   :else
   (s/cond-pre
    ;; literal datetime strings and Java types will get transformed to `absolute-datetime` clauses automatically by
    ;; middleware so drivers don't need to deal with these directly. You only need to worry about handling
    ;; `absolute-datetime` clauses.
    TemporalLiteralString

    #?@(:clj
        [java.time.LocalTime
         java.time.LocalDate
         java.time.LocalDateTime
         java.time.OffsetTime
         java.time.OffsetDateTime
         java.time.ZonedDateTime]

        :cljs
        [js/Date]))))

(def ^:private EqualityComparable
  "Schema for things things that make sense in a `=` or `!=` filter, i.e. things that can be compared for equality."
  (s/maybe
   (s/cond-pre
    s/Bool
    s/Num
    s/Str
    DatetimeLiteral
    FieldOrRelativeDatetime
    ExpressionArg
    value)))

(def ^:private OrderComparable
  "Schema for things that make sense in a filter like `>` or `<`, i.e. things that can be sorted."
  (s/if (partial is-clause? :value)
    value
    (s/cond-pre
     s/Num
     s/Str
     DatetimeLiteral
     ExpressionArg
     FieldOrRelativeDatetime)))
```

They were used for arguments to `<` and the like.

Commit:
Try changing test order

Commit:
Remove mt/with-report-timezone-id nil

Commit:
Add year report timezone tests

Commit:
Rename

Commit:
Rename

Commit:
Remove unused tables from useful-dates

Commit:
Remove useful-dates

Commit:
Update helper-text-strings

Commit:
Tidy

calherries commented:
@nllho, can you help me with the helper-text-strings for this `datetimediff` function?

nllho commented:
@calherries Here's a suggestion :)

```
datetimediff(datetime1, datetime2, unit)

Get the difference between two datetime values (datetime2 minus datetime1) using the specified unit of time.

datetime1, datetime2

The columns or expressions with your datetime values.

unit

Choose from: year, month, week, day, hour, minute, or second.
```

Commit:
Swap order of mt/with-report-timezone-id

Commit:
Change with-report-timezone-id; notify databases after running test

Commit:
Use temp setting for report-timezone instead

Commit:
Update helper-test-strings

dpsutton committed:
Handle literals in `datetimediff-base-base`

calherries commented on src/metabase/driver/postgres.clj:
```diff
@@ -296,6 +300,61 @@
+              (hsql/call :* 12 (helper x y :year))
+              (hsql/call
+               :date_part
+               (hsql/raw (format "'%s'" (name unit)))
```
Thanks

calherries committed:
Update src/metabase/driver/postgres.clj

Co-authored-by: metamben <103100869+metamben@users.noreply.github.com>

Commit:
Simplify postgres second

calherries commented on src/metabase/driver/postgres.clj:
```diff
@@ -296,6 +300,61 @@
+                                    (hsql/call
+                                     :/
+                                     (hsql/call :- b a)
+                                     (case unit :hour 3600 :minute 60 :second 1)))))]
```
Agreed. Update: https://github.com/metabase/metabase/pull/25722/commits/4eacc40a976702ce162944ac9dd1e31e95d418e1

Commit:
Tidy: prefer hx arithmetic functions

Commit:
Drop coercion for string timestamp args for now

qnkhuat commented on src/metabase/driver/postgres.clj:
```diff
@@ -296,6 +300,57 @@
+  [driver [_ x y unit]]
+  (let [x (sql.qp/->honeysql driver x)
+        y (sql.qp/->honeysql driver y)
+        helper
```
should we move this out to an actual `defn`?

Should at least be more readable and hopefully easier to have a docstring.

calherries replied:
https://github.com/metabase/metabase/commit/f4077e325f2991cf151d2e261655a17601a25d19

qnkhuat commented on src/metabase/driver/postgres.clj:
```diff
@@ -296,6 +300,57 @@
+        (fn helper [x y unit]
+          (case unit
+            (:year :day)
+            (hx/cast
+             :integer
```
nit
```suggestion
            (hx/->integer
```

the `hx/->integer` will not cast if the function is already an `integer`. I guess it's not needed in this case but it'd be better if we keep using them.

Also, it saves 1 line of code.

qnkhuat commented on src/metabase/driver/postgres.clj:
```diff
@@ -296,6 +300,57 @@
+              (hx/literal (name unit))
+              (hsql/call
+               (case unit :year :age :day :-)
+               (hsql/call :date_trunc (hsql/raw "'day'") y)
```
```suggestion
               (hsql/call :date_trunc (hx/literal :day)  y)
```
I think this will work, we used this in https://github.com/metabase/metabase/blob/deep-copy-dashboards/src/metabase/driver/postgres.clj#L243

calherries replied:
It would work, but I question the value of adding more indirection here. We're already writing in raw SQL with `hsql/call :date_trunc` etc, so the closer everything is to raw SQL the more consistent it reads IMO.

calherries replied:
Using helpers like `hx/+`, `hx/floor` is helpful because the indirection actually makes the code read more similar to raw SQL, but `hx/literal` doesn't achieve that.

qnkhuat replied:
This is more or less a matter of taste, but IMO looking at sth with a keyword is more readable, compare to `"'day'"`, the combination of double-quote and single-quote is a bit verbose I think :D.

calherries replied:
Yeah, either taste or my lack of familiarity. I'll leave it as is for now. I actually prefer the double quotation marks

Commit:
Extract helper

Commit:
Revert "Change with-report-timezone-id; notify databases after running test"

This reverts commit 7abb543bed13d9d13295d0e00ad1293a40e40d53.

Commit:
Use ->timestamptz

Commit:
Use hx/->timestamp

Commit:
Rename dataset

Commit:
Undo changes to datetime-arithmetics? as these will not match by default

qnkhuat commented on test/metabase/query_processor_test/date_time_zone_functions_test.clj:
```diff
@@ -256,3 +256,198 @@
+        (testing "a day"
+          (mt/with-temporary-setting-values [driver/report-timezone "Atlantic/Cape_Verde"] ; UTC-1 all year
+            (is (partial= {:second 86400 :minute 1440 :hour 24 :day 1}
+                          (diffs #t "2022-10-02T01:00:00Z[+01:00]"     ; 2022-10-01T23:00:00Z[-01:00]
```
```suggestion
                          (diffs #t "2022-10-02T01:00:00[+01:00]" 
```
I'm a bit confused by the `Z` since it means UTC, but the offset is +01.
Pretty sure we'll omit it before set SQL parameter, but we should remove it in the test.

calherries replied:
You're right, this format is just an extra verbose way to write a zoned date time. I've shortened them now.

Commit:
Use proper format for offset datetime literals

Commit:
Shorten datetime literals in tests

Commit:
Removing notify-all-databases-updated from report-timezone setting

Commit:
Fix datetime-arithmetics test

Commit:
Revert "Fix datetime-arithmetics test"

This reverts commit 9141582508170bdbe99c4aa74bc4444e41475be6.

Commit:
Revert "Undo changes to datetime-arithmetics? as these will not match by default"

This reverts commit 9cb05f5475b24a78cb89601c918f3440606cc525.

Commit:
Add UTC timezone to tests

Commit:
Coerce strings to datetimes for ISO formats

calherries commented:
@nllho Apologies for not realising this earlier, but I think we need to expand the helper text. There could be confusion between the behaviour of this function compared to other platforms.

This function returns the number of "whole" units between two datetimes. But on other BI platforms (e.g. [Looker](https://support.google.com/looker-studio/answer/9698749?hl=en&ref_topic=7583292)) and databases a similarly named function would return the number of _boundaries_ between two datetimes.

I've added a small explanation for how it works to the PR description. Can you help simplify it for users?

Commit:
Revert "Removing notify-all-databases-updated from report-timezone setting"

This reverts commit 37356435a2e3981deca9ae76118d857babb2aada.

Commit:
Add comment to postgres driver implementation

Commit:
Formatting

Commit:
Override redshift driver/database-supports?

metamben reviewed (approved):
LGTM

Commit:
Fix comments mixed up by refactoring

qnkhuat reviewed (approved):
This LGTM in general. Well done with using DateTime literal for the tests.

calherries commented on shared/src/metabase/mbql/util.cljc:
```diff
@@ -455,7 +455,10 @@
   "Is a given artihmetics clause operating on datetimes?"
```
Agreed, it's misleading. So is the name `datetime-arithmetics?`. After this PR is merged I want to dive into this, because it's usage in the implementation of `:+` is tangled with its meaning.
`https://github.com/metabase/metabase/blob/29299781da66978a313a9452e459eb12d245ab0e/src/metabase/driver/sql/query_processor.clj#L438`

calherries commented on test/metabase/query_processor_test/date_time_zone_functions_test.clj:
```diff
@@ -256,3 +256,206 @@
+          (mt/with-temporary-setting-values [driver/report-timezone "UTC"]
+            (is (partial= {:second 86400 :minute 1440 :hour 24 :day 1}
+                          (diffs "2022-10-02T01:00:00+01:00"      ; 2022-10-02T00:00:00+00:00
+                                 "2022-10-03T00:00:00+00:00"))))) ; 2022-10-02T23:00:00-01:00
```
Jesus. I did some Paredit refactoring and the comments got mixed up, sticking to the outer `)`. Hopefully this clarifies things: https://github.com/metabase/metabase/commit/29299781da66978a313a9452e459eb12d245ab0e

Commit:
Fix comments mixed up by refactoring 2

Commit:
Tidy comment

Commit:
Rename datetimediff to datetime-diff/datetimeDiff

Commit:
Merge branch 'master' into datediff-functions

Commit:
date-add -> datetime-add

Commit:
Linting

qnkhuat commented on shared/src/metabase/mbql/util.cljc:
```diff
@@ -462,7 +462,10 @@
+    #{:datetime-diff :temporal-extract}
+    false
+
+    #{:interval :relative-datetime :datetime-add}
```
could you please add `datetime-subtract` to this set as well?

calherries replied:
@dpsutton @qnkhuat I discovered some issues with this `datetime-arithmetics?` function, and I think this implementation complicates them further. I've fixed the issues separately in [this PR](https://github.com/metabase/metabase/pull/26279). If approved, we can merge those changes into this PR.

dpsutton committed:
add datetime-subtract to `datetime-arithmetics?`

rework tests a bit as well.

qnkhuat commented on shared/src/metabase/mbql/schema.cljc:
```diff
@@ -94,6 +94,12 @@
     "temporal-extract-units"))
 
+(def DatetimeDiffUnits
+  "Valid units to extract from a temporal."
```
this docstring seems off

qnkhuat commented on test/metabase/query_processor_test/date_time_zone_functions_test.clj:
```diff
@@ -319,3 +319,206 @@
+                                   [:expression "datediff2-add"]]})))))))))
+
+(deftest datetime-diff-type-test
+  (mt/test-drivers #{:bigquery-cloud-sdk}
```
are there any reasons this is unique for `bigquery`?

calherries replied:
After some investigation it seems possible to create a `temporal-type`-like multimethod for postgres and mysql by adding another layer of indirection above `hx/type-info`. I feel it's best to implement these in another PR at this stage.

camsaul replied:
You don't need to create a multimethod. You can use `hx/type-info` and then `hx/type-info->db-type` and then `sql-jdbc.sync/database-type->base-type ` and use the Metabase type hierarchy to make sure it derives from` :type/Date` or `:type/DateTime`. It would be easy to write a function to do that.

camsaul replied:
I agree with @qnkhuat that we should be validating this stuff in Clojure-land for all drivers so we don't have to wait for a round trip to see a query failure -- and so we can get meaningful error messages that make it clear the issue is with the MBQL query and not a bug in the QP code

calherries replied:
I've now changed the type validation so it's handled uniformly across all three drivers 
Commits are https://github.com/metabase/metabase/pull/25722/commits/270c042fa6a7d915e42ae47d03308899f09f9be4 and https://github.com/metabase/metabase/pull/25722/commits/3488d1fc28ef64362aa933491aeba606590d423d

Commit:
Update docstring

Commit:
Use ->temporal-type and trunc to handle report-timezone for bigquery

calherries committed:
Merge branch 'master' into datediff-functions

noahmoss committed:
Log errors caught during sync steps (#26306)

* log errors caught during sync steps

* remove accidental extra parens

calherries committed:
Errors combining datetime interval addition with datetime functions (#26279)

* Add failing tests

* Fix failing tests

* Update shared/src/metabase/mbql/util.cljc

Co-authored-by: Ngoc Khuat <qn.khuat@gmail.com>

* datetime-add and datetime subtract should annotate type by col type

* Fix infer-expression-type for datetime-add/subtract with second, minute, hour

* Undo last commit; they actually always return :type/DateTime

* Fix test based on last commit

* Undo unrelated refactor

* Only test drivers that support expressions

* Only test drivers that support expressions, again

* Update tests from legacy mbql

* Change infered-col-type to be a function again, not macro

* Fix test

Co-authored-by: Ngoc Khuat <qn.khuat@gmail.com>

Commit:
whitespace

Commit:
Remove comments

Commit:
Add explanation for datetime_diff

Commit:
Refactor: replace cast and add `mt/with-driver :bigquery-cloud-sdk` where report-timezone is relevant

Commit:
Add failing tests

Commit:
Fix failing tests

Commit:
Technically it should be bigquery-type

Commit:
Merge branch 'cal-bigquery-timestamp-conversion' into datediff-functions

calherries committed:
Merge branch 'master' into datediff-functions

camsaul commented on deps.edn:
```diff
@@ -114,6 +114,7 @@
   org.clojure/math.combinatorics            {:mvn/version "0.1.6"}              ; combinatorics functions
   org.clojure/math.numeric-tower            {:mvn/version "0.0.5"}              ; math functions like `ceil`
   org.clojure/tools.logging                 {:mvn/version "1.2.4"}              ; logging framework
+  org.clojure/tools.macro                   {:mvn/version "0.1.5"}              ; local macros
```
I can't find anywhere where you're using this

dpsutton replied:
i had it in one part of the change but its no longer there. This is however a transitive dep brought in by compojure so we should leave it in. We use it in a few places

```shell
compojure/compojure 1.6.2
  . org.clojure/tools.macro 0.1.5
```

camsaul commented on frontend/src/metabase-lib/expressions/config.js:
```diff
@@ -367,6 +367,12 @@ export const MBQL_CLAUSES = {
+  "datetime-diff": {
+    displayName: `datetimeDiff`,
+    type: "number",
+    args: ["expression", "expression", "string"],
```
this is not really part of this PR but does this mean you the FE lets you put non-temporal stuff in there? Maybe the expression editor should have the notion of arithmetic expressions vs temporal expressions vs string expressions etc.

camsaul commented on modules/drivers/bigquery-cloud-sdk/src/metabase/driver/bigquery_cloud_sdk.clj:
```diff
@@ -316,6 +316,9 @@
 
 (defmethod driver/supports? [:bigquery-cloud-sdk :foreign-keys] [_ _] true)
 
+(defmethod driver/database-supports? [:bigquery-cloud-sdk :datetime-diff]
+  [_driver _feat _db] true)
```
this is in the bad examples here https://guide.clojure.style/#one-line-functions

```suggestion
(defmethod driver/database-supports? [:bigquery-cloud-sdk :datetime-diff]
  [_driver _feat _db]
  true)
```

calherries replied:
Updated in https://github.com/metabase/metabase/commit/aaa8ac665d597cc1f1d8e2485d3d59050e6c4b56

camsaul commented on test/metabase/query_processor_test/date_time_zone_functions_test.clj:
```diff
@@ -319,3 +319,206 @@
+                                   [:expression "datediff2-add"]]})))))))))
+
+(deftest datetime-diff-type-test
+  (mt/test-drivers #{:bigquery-cloud-sdk}
```
```suggestion
  (mt/test-drivers (mt/normal-drivers-with-feature :datetime-diff)
```

calherries replied:
I plan to implement this in a subsequent PR. Is that OK?

camsaul commented on test/metabase/query_processor_test/date_time_zone_functions_test.clj:
```diff
@@ -319,3 +319,206 @@
+  (mt/test-drivers (mt/normal-drivers-with-feature :datetime-diff)
+    (mt/dataset sample-dataset
+      (testing "Args can be expressions that return datetime values"
+        (let [diffs (fn [x y]
+                      (let [units [:second :minute :hour :day :week :month :year]]
+                        (->> (mt/run-mbql-query orders
+                               {:limit 1
+                                :expressions (into {} (for [unit units]
+                                                        [(name unit) [:datetime-diff x y unit]]))
+                                :fields (into [] (for [unit units]
+                                                   [:expression (name unit)]))})
+                             mt/rows first
+                             (zipmap units))))]
+          (is (= {:second 31795200, :minute 529920, :hour 8832, :day 368, :week 52, :month 12, :year 1}
+                 (diffs [:datetime-add #t "2022-10-03T00:00:00" 1 "day"] [:datetime-add #t "2023-10-03T00:00:00" 4 "day"])))))
```
Instead of running all the queries, sticking the results in a map, and doing a single `(is (= ...))`, I'd suggest taking that map and using `doseq` and doing ones assertion for each. e.g. something like

```clj
(doseq [[unit expected] {:second 31795200, :minute 529920, :hour 8832, :day 368, :week 52, :month 12, :year 1}]
  (testing unit
    (is (= expected
           (mt/run-mbql-query ...)))))
```

it's easier to understand the test output that way especially if one of the queries fails with an Exception or something

calherries replied:
Doing it that way would run a query per unit, which slows down test runs significantly for bigquery. The tradeoff was worth it for development productivity IMO, which matters more than the gain in readability for errors in CI.

camsaul commented on src/metabase/util/honeysql_extensions.clj:
```diff
@@ -293,6 +293,7 @@
 (defn ->date                     "CAST `x` to a `date`."                     [x] (maybe-cast :date x))
 (defn ->datetime                 "CAST `x` to a `datetime`."                 [x] (maybe-cast :datetime x))
 (defn ->timestamp                "CAST `x` to a `timestamp`."                [x] (maybe-cast :timestamp x))
+(defn ->timestamptz              "CAST `x` to a `timestamptz`."              [x] (maybe-cast :timestamptz x))
```
I think only Postgres has a type called `timestamptz` so I don't think this really makes sense inside a general namespace like this for use in a bunch of different places. It seems like it belongs in a Postgres-specific namespace

calherries replied:
Updated in https://github.com/metabase/metabase/pull/25722/commits/4a5c1eecb488a8beffcf31b16247356998d6ae6c

camsaul commented on src/metabase/query_processor/middleware/auto_bucket_datetimes.clj:
```diff
@@ -71,6 +71,8 @@
    ;; do not auto-bucket fields inside a `:time-interval` filter -- it already supplies its own unit
    (mbql.u/is-clause? :time-interval x)
+   ;; do not auto-bucket fields inside a `:datetime-diff` clause: the precise timestamp is needed for the difference
+   (mbql.u/is-clause? :datetime-diff x)
```
`is-clause?` can check multiple clauses at once like `#{:time-interval :datetime-diff}` so you could combine these into one function call if you wanted. Not that it makes a big difference but it is slightly more concise

camsaul commented on src/metabase/driver/postgres.clj:
```diff
@@ -298,6 +302,73 @@
+                    {:valid-units [:year :month :week :day :hour :minute :second]
+                     :bad-unit unit}))))
+
+;; x : date/datetime
+;; y : date/datetime
+;; unit : year | month | week | day | hour | minute | second
+;;
+;; The result is positive if x <= y, and negative otherwise.
+;;
+;; Days, weeks, months, and years are only counted if they are whole to the "day".
+;; For example, `datetimeDiff("2022-01-30", "2022-02-28", "month")` returns 0 months.
+;;
+;; If the values are datetimes, the time doesn't matter for these units.
+;; For example, `datetimeDiff("2022-01-01T09:00:00", "2022-01-02T08:00:00", "day")` returns 1 day even though it is less than 24 hours.
+;;
+;; Hours, minutes, and seconds are only counted if they are whole.
+;; For example, datetimeDiff("2022-01-01T01:00:30", "2022-01-01T02:00:29", "hour") returns 0 hours.
+(defmethod sql.qp/->honeysql [:postgres :datetime-diff]
```
I like this documentation but this is the wrong place to put it. You should put this documentation in `metabase.mbql.schema`

calherries replied:
Updated in https://github.com/metabase/metabase/commit/2311e5642bf90a93eec9d81a960aed76e046e53d

camsaul commented on src/metabase/driver/postgres.clj:
```diff
@@ -298,6 +302,73 @@
+;; For example, datetimeDiff("2022-01-01T01:00:30", "2022-01-01T02:00:29", "hour") returns 0 hours.
+(defmethod sql.qp/->honeysql [:postgres :datetime-diff]
+  [driver [_ x y unit]]
+  (let [x (sql.qp/->honeysql driver (cond-> x
+                                      (string? x) u.date/parse))
```
I think it's a sign that something is wrong if we're having the driver code parse timestamp string literals inside the clauses themselves. This should already be done at this point -- QP middleware should have taken care of it. It's better to handle this stuff once in QP middleware than to have to write the same code 20 times for 20 different drivers to handle it.

I think what you probably want to do is add some code in the `wrap-value-literals/wrap-value-literals-in-mbql` to handle `:datetime-diff` and write a test just for that middleware to make sure it's doing the right thing

camsaul commented on src/metabase/driver/postgres.clj:
```diff
@@ -298,6 +302,73 @@
+       (hsql/raw "'month'")
+       (hsql/call
+        :age
+        (hsql/call :date_trunc (hsql/raw "'day'") y)
```
Why don't you use the `date-trunc` function we already have in this namespace?

camsaul replied:
`->timestamp` in this namespace uses `hx/cast-unless-type-in` and only adds a cast for things that aren't already a `timestamp`, `timestamptz`, or `date`.

```clj
(defn- ->timestamp [honeysql-form]
  (hx/cast-unless-type-in "timestamp" #{"timestamp" "timestamptz" "date"} honeysql-form))
```

So this shouldn't be adding any casts unless your `x` and `y` actually need it or you forgot to propagate that information somewhere.

I'm going to push back on your trying to avoid indirection here. It's better if we reuse code as much as possible for stuff like this. For example maybe we have some issue where we're doing unneeded casting in Postgres -- if that's handled by one function, then we just need to fix one function. Having a set of well-tested low-level stuff like `->timestamp` or `date-trunc` makes the higher-level stuff easier to write since you don't need to figure out the nuances and gotchas for stuff like that

calherries replied:
I see what you mean by the benefits of reuse, e.g. from [this change](https://github.com/metabase/metabase/commit/f31221d88831c2cd866479437cc5693d5ea67078) to date-trunc.

I've changed it in https://github.com/metabase/metabase/commit/71f2e30f0df16fe84a9eb932096ffd7a75a3af7c

camsaul commented on src/metabase/driver/postgres.clj:
```diff
@@ -298,6 +302,73 @@
+(defn- datetime-diff-helper [x y unit]
+  (case unit
+    (:year :day)
+    (hx/cast
+     :integer
+     (hsql/call
+      :date_part
```
Would the `extract-integer` function we use for the `sql.qp/date` implementations in this namespace work here? If so, it seems like you could either use that or leverage the existing date extraction implementations we have for `sql.qp/date` instead of reinventing the wheel and duplicating a lot of that functionality

camsaul commented on src/metabase/driver/postgres.clj:
```diff
@@ -298,6 +302,73 @@
+
+    (throw (ex-info (tru "Invalid datetime-diff unit: {0}" unit)
+                    {:valid-units [:year :month :week :day :hour :minute :second]
+                     :bad-unit unit}))))
```
this is missing `qp.error-type/` info -- add a `:type` key with `qp.error-type/invalid-query` or something

calherries replied:
Updated in https://github.com/metabase/metabase/commit/8b90edd4e33f7c4990af9e5b5baaf3a0550db851

camsaul commented on test/metabase/query_processor_test/date_time_zone_functions_test.clj:
```diff
@@ -319,3 +319,206 @@
         (testing title
           (is (= (set expected) (set (test-datetime-math query)))))))))
+
+(mt/defdataset temporal-dataset
+  [["datediff-mixed-types"
+    [{:field-name "index" :base-type :type/Integer}
+     {:field-name "description" :base-type :type/Text}
+     {:field-name "tz" :base-type :type/DateTimeWithTZ}
+     {:field-name "d" :base-type :type/Date}
+     {:field-name "dt" :base-type :type/DateTime}]
+    [[1 "simple comparing across types" #t "2021-08-03T08:09:10.582Z" #t "2022-09-04" #t "2022-10-05 09:19:09"]]]])
```
If you just want to compare different temporal types like `timestamp` vs `datetime` vs `date` vs `timestamptz`, we have the existing `attempted-murders` dataset that has basically one of everything in it -- you can use that instead of creating a one-off dataset here

calherries replied:
Updated in https://github.com/metabase/metabase/commit/1b9f0ab9eb3bfdded2ae976fa7ff6e64fa548791

camsaul reviewed (changes requested):
It looks pretty close but I left a handful of notes. Ping me again after making tweaks and I'll re-review

- New MBQL clauses should be documented in `metabase.mbql.schema`, not in `metabase.driver.postgres`
- Can the Postgres impl reuse any of the `sql.qp/date` methods?
- It would be extremely nice to validate the arguments in Clojure-land for Postgres and MySQL and add tests for that.
- Individual driver impls should not all have to write their own code to parse datetime literal strings into something they can work with. That should be done once generally by QP middleware instead of 3 times for 3 drivers or 20 times for 20 drivers in the future... `wrap-value-literals` is the right place I think.
- Other small notes

camsaul commented on shared/src/metabase/mbql/normalize.cljc:
```diff
@@ -169,6 +169,13 @@
     [:temporal-extract (normalize-tokens field :ignore-path) (maybe-normalize-token unit) (maybe-normalize-token mode)]
     [:temporal-extract (normalize-tokens field :ignore-path) (maybe-normalize-token unit)]))
 
+(defmethod normalize-mbql-clause-tokens :datetime-diff
+  [[_ x y unit]]
+  [:datetime-diff
+   (normalize-tokens x :ignore-path)
+   (normalize-tokens y :ignore-path)
+   (maybe-normalize-token unit)])
```
Did you add tests for this?

calherries replied:
Now I did https://github.com/metabase/metabase/pull/25722/commits/e389aaf3fbfb6d434fa5a9aa8df89c7eeaa4dcc0

camsaul commented on shared/src/metabase/mbql/schema.cljc:
```diff
@@ -94,6 +94,12 @@
+  "Valid units for a datetime-diff clause."
+  (s/named
+    (apply s/enum #{:second :minute :hour :day :week :month :year})
+    "temporal-extract-units"))
```
💀

camsaul replied:
```suggestion
    "datetime-diff-units"))
```

camsaul commented on shared/src/metabase/mbql/schema.cljc:
```diff
@@ -476,7 +482,8 @@
   "Set of valid arithmetic expression clause keywords."
-  #{:+ :- :/ :* :coalesce :length :round :ceil :floor :abs :power :sqrt :log :exp :case})
+  #{:+ :- :/ :* :coalesce :length :round :ceil :floor :abs :power :sqrt :log :exp :case
+    :datetime-diff})
```
I saw that you had a question somewhere (on this PR I think) about whether `arithmetic-expressions` meant expressions whose arguments are numeric, or expressions whose results are numeric, and you got an answer. Please write that sort of stuff down in comments, or add it to the docstring! That way the next person who has that question will have the answer instead of them having to do the same thing you did

calherries replied:
Updated in https://github.com/metabase/metabase/pull/25722/commits/b9b0ef41fabd438c80f6fb310c50a88aee2e85c3. I've also left a bunch of TODOs in this commit https://github.com/metabase/metabase/pull/25722/commits/b9b0ef41fabd438c80f6fb310c50a88aee2e85c3 because I think we should rename `arithmetic-expressions` to `numeric-expressions` for clarity. There are also more functions that should be in `numeric/arithmetic-expressions` that aren't, e.g. temporal extraction functions like `get-year`. I'll leave it for a separate PR.

calherries committed:
Update modules/drivers/bigquery-cloud-sdk/src/metabase/driver/bigquery_cloud_sdk.clj

Co-authored-by: Cam Saul <1455846+camsaul@users.noreply.github.com>

Commit:
Move documentation to metabase.mbql.schema

calherries committed:
Update DatetimeDiffUnits

Co-authored-by: Cam Saul <1455846+camsaul@users.noreply.github.com>

Commit:
Add error type and optimize case expressions to driver implementations

Commit:
refactor for brevity

Commit:
Fix error with postgres

Commit:
Fix error with postgres

Commit:
Handle string literal parsing in wrap-value-literals

Commit:
Remove ->timestamptz

Commit:
Add comments + TODOs showing arithmetic expressions should return numeric values

Commit:
Add explanation of arithmetic expression as docstring

Commit:
Add test for normalize-mbql-clause-tokens

Commit:
Remove unused import

Commit:
Remove unused form

Commit:
Switch tests to use attempted-murders dataset

Commit:
Undo optimization that broke tests

Commit:
Validate non-temporal types for bigquery

Commit:
Add error handling for incorrect types

Commit:
Fix mysql type checking

Commit:
Fix mysql type checking

camsaul commented on src/metabase/driver/mysql.clj:
```diff
@@ -348,6 +353,40 @@
+  [driver [_ x y unit :as clause]]
+  (let [x (sql.qp/->honeysql driver x)
+        y (sql.qp/->honeysql driver y)
+        disallowed-types (keep
+                          (fn [v]
+                            (when-let [db-type (some-> v hx/type-info hx/type-info->db-type keyword str/upper-case)]
+                              (let [base-type (sql-jdbc.sync/database-type->base-type driver db-type)]
+                                (when-not (some #(isa? base-type %) [:type/Date :type/DateTime])
+                                  (name db-type)))))
+                          [x y])]
+    (when (seq disallowed-types)
+      (throw (ex-info (tru "Only datetime, timestamp, or date types allowed. Found {0}"
+                           (pr-str disallowed-types))
+                      {:found disallowed-types
+                       :type  qp.error-type/invalid-parameter})))
```
I was thinking in would be nice if this could be abstracted out somewhere so we could reuse it between the different drivers, or at least if some of the stuff implementation stuff could be abstracted out 

For example the type-info -> db-type -> base-type stuff could be a good thing to put in some general util namespace for `:sql-jdbc` drivers. But we actually don't have any appropriate place to put it currently I think. So we'd have to make a new namespace.

I guess it's ok for now unless you want to make some sort of `:sql-jdbc` util namespace to put this in

calherries replied:
Yes, it would be nice to have a kind of static type checker for MBQL. The MBQL schema is already doing some type checking because it knows the argument and return types of some expressions, but it doesn't know the type of fields. Perhaps there's a way to integrate field types into the schema definition.

Regardless, once I've implemented datetime-diff for a few more drivers I'll strip out this repeated stuff into a util namespace like you suggested.

camsaul commented on src/metabase/driver/mysql.clj:
```diff
@@ -348,6 +353,40 @@
+    (when (seq disallowed-types)
+      (throw (ex-info (tru "Only datetime, timestamp, or date types allowed. Found {0}"
+                           (pr-str disallowed-types))
+                      {:found disallowed-types
+                       :type  qp.error-type/invalid-parameter})))
```
This is the definition of `invalid-parameter`

```clj
(deferror invalid-parameter
  "The query is parameterized, and a supplied parameter has an invalid value."
  :parent invalid-query
  :show-in-embeds? true)
```

This means parameter in the Dashboard or SQL `{{template tag}}` parameter sense. It's not the right error type here. 

We're not really doing a lot with the error types at the moment -- the old BigQuery driver had logic to retry on certain query errors like timeouts but would not retry malformed queries -- don't remember if this is still in the new BigQuery driver or not. I don't think we're using them in too many other places. Either way we should still try to use error types that make sense so we can leave the door open to do something useful with that info in the future.


```suggestion
                      {:found disallowed-types
                       :type  qp.error-type/invalid-query})))
```

`invalid-query` is the closest thing we have here that makes sense. You could add a subtype called `type-mismatch` or something if you wanted. I'm cool either way. Just don't use `invalid-parameter` for something that doesn't have to do with query `:parameters`

calherries replied:
Makes sense. Updated in https://github.com/metabase/metabase/commit/6d0485c90dcc2aaff9b50448cbccc98f50f87b19

camsaul commented on src/metabase/driver/mysql.clj:
```diff
@@ -348,6 +353,40 @@
+      (:hour :minute :second)
+      (hsql/call :timestampdiff (hsql/raw (name unit)) x y)
+
+      (throw (ex-info (tru "Invalid datetime-diff unit {0}" unit)
+                      {:clause      clause
+                       :valid-units [:year :month :week :day :hour :minute :second]
+                       :type        qp.error-type/invalid-parameter})))))
```
This is not something that should be happening in each individual driver. If it's in the MBQL schema then the validate-query middleware, which happens right after normalization, will reject it. 

In general when working on QP/driver stuff stop and ask yourself "is this something that can be done once generally, or is it something that has to be done 20 times for 20 drivers?". The only things that need to be done for each individual driver are things that are driver-specific, like how to translate a specific MBQL clause to their flavor of SQL/native query language. MBQL transformations and validation are not driver-specific and should not be done in each individual driver.

This doesn't really apply to the type validation at the beginning of this method since column type info is DB-dependent. Maybe in the future we can have better typing around MBQL stuff and some of that can be determined by middleware before it ever hits the QP but for the time being it's ok

calherries replied:
Yes, this is actually caught by the MBQL schema anyway. I've removed this error handling in https://github.com/metabase/metabase/commit/e2a8f3be32855048d1310163991e4dd0b919f0cb 

In theory the `unit` argument should be allowed to be a string expression or field, and if so, we'd have to bring back this error handling to each driver, right? Either way, I'd leave it for the future.

camsaul commented on src/metabase/driver/postgres.clj:
```diff
@@ -298,6 +303,70 @@
+       :month
+       (hsql/call
+        :age
+        (hsql/call :date_trunc (hsql/raw "'day'") y)
```
```suggestion
        (hsql/call :date_trunc (hx/literal :day) y)
```

might be a little easier on the eyes

calherries replied:
Updated in https://github.com/metabase/metabase/commit/71f2e30f0df16fe84a9eb932096ffd7a75a3af7c

camsaul reviewed (approved):
I left a few more notes but I think this is close enough I'll give it a tentative ✔️ 

You shouldn't need to be checking the `unit` inside the driver methods since presumably the MBQL schema `validation` middleware should reject invalid units -- if it's not rejecting it then it should be 

I would also still prefer trying to reuse more existing code in the `postgres` driver namespace so if there's a bug with something or we just want to optimize things we only have to fix one function instead of a whole bunch of places. 

It seems like adding another test for a `:field` with temporal bucketing applied would be nice, just to get a little more coverage. Presumably bucketing by an extraction unit like `:month-of-year` would fail because that returns an `integer` which isn't allowed here. But if you have two fields and truncate them both to `:month`, then the `:month` difference should be the same as if you didn't bucket them at all... right? I really looked at this code a lot trying to think of any situations where it wouldn't work but I think it all looks good... never hurts to have a few more tests cases tho just to be sure. 

```clj
[:datetime-diff [:field 1 {:temporal-unit :month}] [:field 2 {:temporal-unit :month}] :month]
```

Besides those suggestions everything looks good so feel free to tweak and then merge without waiting for a third review

Commit:
invalid-parameter -> invalid-query

Commit:
Use date-trunc and extract

Commit:
Remove unit error handling

calherries committed:
Merge branch 'master' into datediff-functions

Commit:
DatetimeLiteral -> DateOrDatetimeLiteral

Commit:
Remove unused binding

Commit:
Fix mysql type checking

calherries commented:
@camsaul 
I don't know if `[:datetime-diff [:field 1 {:temporal-unit :month}] [:field 2 {:temporal-unit :month}] :month]` is possible to generate from the GUI editor. It doesn't make sense to time-bucket a field and use the result in a custom expression, so I assume it will never be possible.

Even if it were possible in the future, `[:datetime-diff [:field 1 {:temporal-unit :month}] [:field 2 {:temporal-unit :month}] :month]` wouldn't always return the same result as `[:datetime-diff [:field 1 nil] [:field 2 nil] :month]`. `datetime-diff` is meant to return whole months, not part months. See the explanation in the comments: https://github.com/metabase/metabase/blob/e98b3bd59c9d15c22d9d2663262360e44d8ededa/shared/src/metabase/mbql/schema.cljc#L676

calherries committed:
Merge branch 'master' into datediff-functions
